File: KalkHead/binarytree.cpp
#include "binarytree.h"

void BinaryTree::distruggi(nodo* t){
    if(t){
        distruggi(t->left);
        distruggi(t->right);
        delete t;
    }
}

int BinaryTree::conta(nodo* t){
    if(t)
        return 1 + conta(t->left) + conta(t->right);
    return 0;
}


BinaryTree& BinaryTree::operator =(const BinaryTree& t){
    if(this != &t){
        distruggi(root);
        root = copia(t.root);
    }
    return *this;
}

BinaryTree::nodo::~nodo(){
    if(info)
        delete info;
}

bool BinaryTree::empty() const{

    if(root == nullptr){
        std::cout << "INSIDE EMPTY" << std::endl;
        return true;
    }
    else{
        return false;
    }
}
//PER GUI

int BinaryTree::max_depth() const{
    return height(root);
}

int BinaryTree::height(nodo *n) const{
    if (!n)
        return -1;
    return 1 + std::max(height(n->left), height(n->right));
}

BinaryTree::nodo* BinaryTree::returnRoot() const{
    return root;
}

void BinaryTree::elimTree(){
    if(root){
        delete root;
        root = nullptr;
    }
}

File: Java/src/BinaryTree.java
public abstract class BinaryTree {

    // Perche static ?
    protected static class Nodo{
        protected Tipo info;

        protected Nodo left;
        protected Nodo right;
        protected Nodo parent;

        protected Nodo(Tipo t, Nodo p, Nodo l, Nodo r){
            info = t;
            parent = p;
            left = l;
            right = r;
        }

        protected Nodo(Tipo t, Nodo p, Nodo l){
            this(t, p, l, null);
        }

        protected  Nodo(Tipo t, Nodo p){
            this(t, p, null, null);
        }

        protected Nodo(Tipo t){
            this(t, null, null, null);
        }

        protected Nodo(){
            this(null, null, null, null);
        }

    }



    public abstract void insert(Tipo t);
    public abstract Tipo search(Tipo t);
    public abstract Tipo max();
    public abstract Tipo min();

    public abstract BinaryTree copia();

    public abstract void deleteNodo(Tipo t);

    public boolean empty(){
        if(root == null)
            return true;
        else
            return false;
    }

    public abstract void print();

    protected Nodo root = null;

    protected abstract Nodo copiaRoot(Nodo t, Nodo p);

    // non serve il metodo per distruggere
    protected static int conta(Nodo r){
        if(r != null)
            return 1 + conta(r.left) + conta(r.right);
        return 0;
    }

}

